Measure invalidation performance and fixes

albyrock87 opened the pull request:
### Description of Change

Initial proposal to solve #24551 .

I executed a speedscope on [davidortinau/AllTheLists](https://github.com/davidortinau/AllTheLists/blob/net9/src/Pages/LearningPage.xaml#L20) `LearningPage` using the latest nightly `8.0.99-ci.net8.24468.1` build.

What I noticed is that #23052 `OnChildMeasureInvalidatedInternal` was showing up and taking 2% of total time (recursive calls).

What I've done:
- Avoids triggering measure invalidated completely while building the MAUI tree in memory (`Handler = null`)
- Fixes some iOS `SetNeedsLayout` propagation issues and cleans up detection of auto `SetNeedsLayout` parent propagation with `IPropagatesSetNeedsLayout` internal interface
- Removes "out-of-layout-pass" measurements from legacy layouts making them fast as new layouts

### WinUI Speedscope

Kindly provided by @MartyIX

**Before `main`**
![image](https://github.com/user-attachments/assets/0ebb68b3-f1ff-46c6-95c6-9a518e207ae1)
[1726819823.MAIN.speedscope.json](https://github.com/user-attachments/files/17073463/1726819823.MAIN.speedscope.json)

**After `PR`**
![image](https://github.com/user-attachments/assets/cc3fe7fa-a820-4404-81d3-3a7369bbbe37)
[1726819948.PR.speedscope.json](https://github.com/user-attachments/files/17073469/1726819948.PR.speedscope.json)


**Before `main`**
![image](https://github.com/user-attachments/assets/79d83532-489b-438d-9b18-1de39355f096)
[1726819865.MAIN.speedscope.json](https://github.com/user-attachments/files/17073484/1726819865.MAIN.speedscope.json)


**After `PR`**
![image](https://github.com/user-attachments/assets/79ff2cc0-2d75-40de-b7db-c02fcfc9a3be)
[1726819973.PR.speedscope.json](https://github.com/user-attachments/files/17073488/1726819973.PR.speedscope.json)


### iOS Speedscope

**Before `main`**
![image](https://github.com/user-attachments/assets/c90f1ac7-3037-4f1b-a63e-0e522a6f9b5e)
[before.speedscope.json.zip](https://github.com/user-attachments/files/17048820/before.speedscope.json.zip)

**After `PR`**
![image](https://github.com/user-attachments/assets/5fb87751-f90b-4a2c-ad30-02d5b6246c54)
[after.speedscope.json.zip](https://github.com/user-attachments/files/17049097/after.speedscope.json.zip)

### Issues Fixed

Fixes  #24551

bcaceiro commented:
@albyrock87 I'm quite interested on your phrase: " due to legacy layouts (which includes ContentView)" , the example has a CollectionView with a DataTemplate , that has a ContentView as a parent - should we have better performance, in having for instance, DataTemplates with a Grid / Border as a parent?

rmarinho commented:
/azp run

albyrock87 commented:
@bcaceiro after this PR it shouldn't matter if you use ContentView or not.
Before this PR I can say there is a performance advantage in completely avoid views based on legacy Layout class.

albyrock87 committed:
Improve measure invalidation and legacy Layout(s) performance
- Do not bubble up `MeasureInvalidated` on pages
- Invoke `MeasureInvalidated` just once when `BindingContext` is changing
- Do not synchronously measure/arrange children on legacy layout upon child measure invalidated: simply wait for the next native layout pass

albyrock87 commented on src/Controls/src/Core/LegacyLayouts/Layout.cs:
```diff
@@ -192,8 +192,12 @@ private protected override IList<Element> LogicalChildrenInternalBackingStore
+			var request = new Size(size.Request.Width + Padding.HorizontalThickness, size.Request.Height + Padding.VerticalThickness);
+			var minimum = new Size(size.Minimum.Width + Padding.HorizontalThickness, size.Minimum.Height + Padding.VerticalThickness);
+
+			DesiredSize = request;
```
Base method sets `DesiredSize` so the fact this was missing was simply wrong.

albyrock87 commented on src/Controls/src/Core/Page/Page.cs:
```diff
@@ -583,29 +582,6 @@ protected void UpdateChildrenLayout()
-				}
-			}
-
-			InvalidateMeasureInternal(InvalidationTrigger.MeasureChanged);
```
This was already not happening due to the fact there was no `MeasureInvalidated` propagation prior #23052

PureWeen replied:
No :-/ I've been wanting to get rid of this code

The main edge case I could see is if someone was using shimmed pages on iOS but in theory those should still follow the maui layout system via the shim wrapper

PureWeen commented:
/azp run

PureWeen commented:
/azp run

PureWeen commented:
/azp run

albyrock87 committed:
Add a UI test to log rendering performance

albyrock87 committed:
Remove `ApplyBindings` and reduce `OnChildMeasureInvalidatedInternal` switch branches

albyrock87 commented on src/Controls/src/Core/VisualElement/VisualElement.cs:
```diff
@@ -1354,10 +1374,38 @@ public void InvalidateMeasureNonVirtual(InvalidationTrigger trigger)
+		}
 
+		bool _isApplyingBindings;
+		InvalidationTrigger? _applyingBindingsInvalidationTrigger;
```
It seems like the additional memory usage is worth it from my performance tests.
I had to revert this part per @PureWeen request, but I'll think about your feedback on the follow-up PR.
We can always change `InvalidationTrigger` to a `enum InvalidationTrigger: byte`.. it's a public API enum, but is used only on a `IVisualElementController` public method.

jonathanpeppers reviewed (dismissed):
I don't really know the details of how MAUI's layout works, but this seems ok to me.

The fields I was asking about are removed now. 👍

PureWeen committed:
Update Issue24532.xaml.cs

PureWeen commented:
/azp run

albyrock87 commented on src/Controls/src/Core/LegacyLayouts/Layout.cs:
```diff
@@ -517,20 +527,13 @@ internal virtual void OnChildMeasureInvalidated(VisualElement child, Invalidatio
-				if (trigger == InvalidationTrigger.HorizontalOptionsChanged || trigger == InvalidationTrigger.VerticalOptionsChanged)
+				if (trigger is InvalidationTrigger.HorizontalOptionsChanged or InvalidationTrigger.VerticalOptionsChanged)
 				{
 					ComputeConstraintForView(view);
```
I just realized something..
`LegacyLayout` has this code which invalidates view legacy constraints with `ComputeConstraintForView`
But now, I'm not propagating MeasureInvalidated on *Options changes.
Which means this code will never run, and potentially generates wrong behavior.

So either:
- This entire method can be completely removed (if nowadays `ComputeContraintForView` is some kind of legacy XF code)
- My choice to not propagate `*Options` invalidation is wrong

albyrock87 commented on src/Controls/src/Core/LegacyLayouts/Layout.cs:
```diff
@@ -503,7 +513,7 @@ internal virtual void OnChildMeasureInvalidated(VisualElement child, Invalidatio
 			for (var index = 0; index < count; index++)
 			{
-				if (LogicalChildrenInternal[index] is VisualElement v && v.IsVisible && (!v.IsPlatformEnabled || !v.IsPlatformStateConsistent))
+				if (LogicalChildrenInternal[index] is VisualElement { IsVisible: true } v && (!v.IsPlatformEnabled || !v.IsPlatformStateConsistent))
```
This line can very likely be removed thanks to the fact that `IsPlatformEnabled` is now being checked on `VisualElement` before triggering measure invalidation.

albyrock87 committed:
Account for view `Constraint` and propagate `InvalidationTrigger` on parent correctly

albyrock87 committed:
Now that we are not invalidating manually each `LegacyLayout`, we have to invalidate parent chain through the platform view on `ScrollView` (iOS)

albyrock87 committed:
Sync behavior on Page

PureWeen commented:
/azp run

albyrock87 committed:
Adds a rendering performance test

albyrock87 commented:
I'd like to wait for #24762 and rebase so we can actually measure performance everywhere :)

albyrock87 committed:
Removes old perf test

albyrock87 committed:
Additional refactor

albyrock87 committed:
Removes performance test

albyrock87 committed:
Remove old comments

albyrock87 committed:
Remove useless branches and fix order of invocation

PureWeen commented:
/azp run

albyrock87 committed:
Fix broken device test

albyrock87 committed:
Third tentative to make things right

albyrock87 committed:
Cleanup and fix enhanced `ViewTests` broken device test

PureWeen commented:
/azp run

albyrock87 commented:
Closing in favor of https://github.com/dotnet/maui/pull/25664
